### ISSUE TITLE
Allow creating a single route out of a relations that has several loops

### ISSUE THREAD

**Issue by Tsurha** [IsraelHikingMap/Site#1807]
## What I expect to happen
When a route relation has several loops, a single long GPX should be created when clicking the + icon in IHM.

## What really happened
Several routes are created :-(

![image](https://user-images.githubusercontent.com/49042192/205353270-09b182c1-2ca9-4755-bd6c-cac550597931.png)

## What I did that caused the issue - step by step
As an example, please open the Gvaraam singletrack information pane: https://israelhiking.osm.org.il/poi/OSM/relation_1373988?language=he
Click the + and create a route.

## Things I think the developers should know, images, links etc.
The solution may include a method or adjustments in creating the relation in OSM. IHM could read that info and create the GPX appropriately. In such case, an explanation of how to do so in OSM is required.

**Comment by HarelM** [comment]
Any chance you can find me a smaller example I can investigate? This is a big of a big route to debug and it would be great to look at a simpler example.

**Comment by Tsurha** [comment]
A classic route topology is in a letter Q shape.  Starting at the tip of the appendage in a parking lot, cycling to the loop, then looping.  My tendency was not include the little entry path in the relation. In the following two its with that chupchik:
- Odem: https://israelhiking.osm.org.il/poi/OSM/relation_11085528?language=he
- Hashluch: https://israelhiking.osm.org.il/poi/OSM/relation_10965978?language=he
- Furidis: https://israelhiking.osm.org.il/poi/OSM/relation_10966432?language=he

This one is supposed to be two loops connected with a line: https://israelhiking.osm.org.il/poi/OSM/relation_13082537?language=he
I did not complete one loop because of the current + algorithm... 

This one is 3 loops: https://israelhiking.osm.org.il/poi/OSM/relation_4192239?language=he

**Comment by HarelM** [comment]
This is basically a duplicate of #1203. I don't remember the reason for not addressing this issue back then besides the fact that it didn't have a high priority.

**Comment by HarelM** [comment]
I'm digging into the algorithm and there are some cases which I'm not sure I understand why I decided to do it the way I did.
Some are described here: #1078, other are described in tests.
But it seems that the first part of the סינגל גברעם is a Q letter which was converted successfully. So, it might be possible to simply change the order of ways to fix some of these issues.
@zstadler what's the easiest way to change the order of ways in a relation?

**Comment by HarelM** [comment]
Almost there... :-(
<img width="1037" alt="image" src="https://user-images.githubusercontent.com/3269297/206399130-960549b0-fb6c-4890-bc9f-51c9f2eccf77.png">
I think the recent changes are an improvement though:
<img width="1037" alt="image" src="https://user-images.githubusercontent.com/3269297/206399596-b4926fe8-0503-4145-8445-cfcd5135c174.png">
<img width="1174" alt="image" src="https://user-images.githubusercontent.com/3269297/206399789-5bf333c5-d73b-4153-8084-41c86b06a29d.png">
I'll push these changes to the site later today so you'll have time to go through the new routes.
I think סינגל גברעם might be a very edge case compared to the others...

**Comment by Tsurha** [comment]
It sure looks much better!!!
This will allow creating better singletrack relations without bothering with workarounds. Thank you!

**Comment by HarelM** [comment]
You can test it, I just deployed it this morning.
Let me know if it cover most cases.

**Comment by Tsurha** [comment]
There is a notable improvement. Certainly in the above examples. **Well done!**

To test more, I added some appendages to a few loops. These show how to get from the parking lot to the singletrack loop. Basically converting an O to a Q.  

### Please have a look at the following:

- Givat Hmore https://israelhiking.osm.org.il/poi/OSM/relation_5828701?language=he Is now a Q that is supposed to start at the gas station’s parking lot. For some reason 2 private routes are created with the +

- Hof Carmel full https://israelhiking.osm.org.il/poi/OSM/relation_8774548?language=he is supposed to start from its Q appendage in the Zeitim Parking but does not. It is also supposed to be counter clockwise.  All the segment arrows are pointed well, but the route is reversed. 

- The loop in Blue Beeri: https://israelhiking.osm.org.il/poi/OSM/relation_10848253?language=he is supposed to have a clockwise direction. However this would contradict shared segments with the Orange and the Green trail. Have a look at the raster map in its picture showing arrows for all three. Therefore it is presented wrongly in IHM and I don't know how to fix it.

### Questions / Improvements

- With a Q route, how can I create the appendage in both directions? It is starting from the parking to the loop and finishing from the loop to the parking.  Both above mentioned Beeri should be that way.

- If a singletrack is a full circle - O. How do I define its starting point? Do I just break it there as I did in Emeq Hammayanot https://israelhiking.osm.org.il/poi/OSM/relation_10984643?language=he. This is not an elegant solution but a workaround that works.

### Other issues for your consideration:

- Yafia Singletrack: https://israelhiking.osm.org.il/map/15.21/32.6982/35.2700 is supposed to have a yellow background in the MTB map, just like its brother Kfar Hachoresh. It does not.

**Comment by HarelM** [comment]
Yes, I thought this Q problem where the ways that are part of a loop won't start from the beginning... Not sure it's worth my time tough.

The O part is a problem which I'm not sure how to solve besides creating a small "gap" in the relation so that it won't be a full circle or ordering the ways using JOSM.

The icons are there, they are just exactly one on top of the other, so you only see one of them when you zoom in:
![Screenshot_20221210-143530_Israel Hiking Map](https://user-images.githubusercontent.com/3269297/206855569-cb866791-32db-4b46-a04b-db5ad4d007d3.jpg)

MTB map rendering is outside the scope of this issue and repo.

Bottom line, I'm not sure OSM is the right place to store bike routes given all these considerations...

**Comment by Tsurha** [comment]
Regarding your bottom line, OSM has its pros and cons:
- Pro: They allow the community to fix issues. 
- Pro: When people adjust segments to better reflect their position, the routes are improved.  
- Con: When they break relations, we have a problem. 
- Con: Representing complex topologies is a problematic.

I'd appreciate if you could at least try to see why didnt Givat Hamore convert well like other Q ones that did

**Comment by HarelM** [comment]
The reason for two ways instead of one is fairly obvious (to me) - the order of the ways. 
The algorithm (scary word, I know) is fairly simple:
1. Get all the ways from the relation (ordered by how they are saved in OSM)
2. Try and group the ways by adding one way to the other (only ways that start or end at the same point), the grouping will create long segments of ways, and the matching is only at the start or the end of a segment (here's the problem basically)
3. If a way has `oneway` or `mtb:oneway` the entire segment the algorithm currently works on will be aligned with the direction of that way (if you have multiple "one-ways", the last one will define the route's direction)
4. Once all ways are processed the geometry is saved, if there is more than one line, you'll see it only when converting to private route, as the POI presentation is a single line for all the geomerty.

What causes the above issue is that the algorithm will find ways that create a circle before trying to group the chupchik of the Q, which will cause a geometry of a circle and a line instead of a single "9" kind of geometry.

I'll see if I can spend a small amount of time to improve the above algorithm short coming, but I can't promise.

**Comment by HarelM** [comment]
After latest fixed, locally still on my PC only:

https://israelhiking.osm.org.il/poi/OSM/relation_5828701?language=he 
![image](https://user-images.githubusercontent.com/3269297/206875498-958b2955-72b8-414f-88c1-20efd88676df.png)

https://israelhiking.osm.org.il/poi/OSM/relation_8774548?language=he
I'm getting a gap which I think you fixed yesterday, but I might not have the latest changes as I took a daily extract.
So I'm optimistic this is also solved.
![image](https://user-images.githubusercontent.com/3269297/206876425-9e489f12-797d-4a3e-bac3-c44738f296a7.png)


https://israelhiking.osm.org.il/poi/OSM/relation_10848253?language=he
![image](https://user-images.githubusercontent.com/3269297/206875606-a11479a9-9e3a-404f-9c70-8c25115e213b.png)

**Comment by Tsurha** [comment]
We have probably worked on different versions. I tried to workaround the current algorithm by creating a gap towards the end of the above 3 examples you gave.  Please let me know once you deployed your last algorithm change and the OSM DB update frequency.  I'll look again.

It is possible to move relation members in ID by dragging. But its difficult to know which is which so I'd rather not use that. 
What is easy to do is to give the first member a role.  This will allow you to start sorting them from the parking lot.  It would reduce the issue you explained above in point 2 of the algorithm.  For you consideration.

Screen capture that shows both a role name and dragging its member higher.

![image](https://user-images.githubusercontent.com/49042192/206901994-e80f89cd-1cb4-450b-9528-07f2b893af80.png)

**Comment by HarelM** [comment]
The updated version is up and running on the server, you can test how it behaves without the gaps and decide which way to go.

**Comment by Tsurha** [comment]
Creating a gap means converting a 9 to a line. So far this works well.

When the 9's loop is full without a gap, a single private route is now created once clicking +, which is an improved behavior. Nice work!

However, I do not know how to control the direction in the 9’s loop. I tried checking point number 3 in the above algorithm explanation, for this one:
https://israelhiking.osm.org.il/poi/OSM/relation_10966432?language=he
It is supposed to have a counterclockwise direction. Could not make that happen. 

For the OSM editor, number 3 in the algorithm means checking all the tags of all segments of a singletrack. Too time consuming in OSM. It would be doable on a list if possible.  We need a better method to set the direction. Seems like what Zeev is asking here: https://github.com/IsraelHikingMap/Site/issues/1813

**Comment by HarelM** [comment]
#1813 is talking about the route that you currently edit, it's not talking about the OSM relations.
Direction is super tricky, and I don't have a good idea how to solve this.
A thing you can try out is the following:
1. Open the relation in IHM
2. Click to see it in OSM
3. Find the last segment in the "loop" in terms of the real last segment (you can do it using the editor)
4. Move it to be the last segment in the relation (as you said can be done above)
5. Mark it with `mtb:oneway`

This does not have the need to reorder all the relation, just find and move the last segment in reality to be the last segment in the OSM relation.

Worth trying it on one route and then decide if this is a solution we would like to propagate to other routes...

If you have other ideas, I'll be happy to hear.
Calculation clockwise and counterclockwise is hard and I would like to avoid it, especially if a user can simply click the "reverse route" button...

BTW, if the gaps are not needed anymore (which I think they are not after the latest fix) I would consider reverting them, but this is just a suggestion.

**Comment by Tsurha** [comment]
Just tried. Locating the segment in a relation is difficult while possible. I managed but not sure if that is a good ongoing work method. How do I mark it with mtb:oneway in ID though?????

![whereis mtb oneway](https://user-images.githubusercontent.com/49042192/207630048-21aeded7-2eeb-461e-aa0d-1fd0a5bd3e35.png)
